lib/features/support/interactive_cli.rb

```diff
@@ -0,0 +1,117 @@
+require 'open3'
+
+# Encapsulates a terminal session, retaining state and input streams for interactive tests
+class InteractiveCLI
+  # @!attribute [r] parsed_output
+  #   @return [Array] An array of output strings received from the terminals STDOUT pipe
+  attr_reader :parsed_output
+
+  # @!attribute [r] parsed_error
+  #   @return [Array] An array of error strings received from the terminals STDERR pipe
+  attr_reader :parsed_errors
+
+  # @!attribute [r] last_exit_code
+  #   @return [Number] The exit code of the last terminal command
+  attr_reader :last_exit_code
+
+  # @!attribute [r] pid
+  #   @return [Number] The PID of the running terminal
+  attr_reader :pid
+
+  # @!attribute [r] current_buffer
+  #   @return [String] A string representation of the current output present in the terminal
+  attr_reader :current_buffer
+
+  # Creates an InteractiveCLI instance
+  #
+  # @param environment [Hash] A hash of environment variables
+  # @param terminal [String] A path to the terminal to run, defaults to `/bin/sh`
+  # @param terminal [String] The stop command, defaults to `exit`
+  def initialize(environment = {}, terminal = '/bin/sh', stop_command = 'exit')
+    @env = environment
+    @parsed_output = []
+    @parsed_errors = []
+    @current_buffer = ''
+    @last_exit_code = nil
+    @in_stream = nil
+    @pid = nil
+    @running = false
+    @stop_command = stop_command
+    start_threaded_terminal(terminal)
```

Comment:
Generally I prefer avoiding doing work in a constructor; could we provide a public `start` method to be called on a new instance instead?

Reply:
I agree, above doing processing beyond basic initialization in constructors - it can be surprising for clients and make testing more difficult.

Reply:
I'm not necessarily sure I agree.  

The purpose of the object is to represent a single-run shell or terminal which can be interacted with.  Adding a single-use `start` method, which must then determine whether the instance has previously been started and handle being called multiple times on a single instance, seems like an unnecessary addition.

Reading the name and purpose of the class I would assume that, by creating an instance of the class, I would be able to interact with a shell, with no further required setup.  You can argue that this can be managed by it being started before it is returned by the `Runner` class, but I'm not sure that actually adds anything other than an extra step.

It may help with unit testing, but even that's easily worked around by stubbing and mocking class methods.

Reply:
My comment was based mainly on experience of similar setups in the past and me having developed a general feeling about it over the years, rather than having a specific concern for this particularly instance.  Certainly opinion and style more than fact in that respect. :-). Also, to be fair, it seems much easier to mock things when unit testing in Ruby than other languages I have dealt with... so if you're happy that you can unit test it effectively without spinning up external processes (for instance), I can many deal with it staying as it is.

Reply:
Yeah I've been bitten by this in the past, but I don't think there's anything especially egregious about this particular case 🙂 — happy for it to stay as-is